README.md
![Screenshot 1](screenshots/screenshot-1.png)

<span align="center">

[![CI](https://github.com/steve228uk/YouTube-Music/actions/workflows/CI.yml/badge.svg)](https://github.com/steve228uk/YouTube-Music/actions/workflows/CI.yml) [![Latest Release](https://img.shields.io/github/v/release/steve228uk/YouTube-Music?color=yellow)](https://github.com/steve228uk/YouTube-Music/releases) [![macOS 10.11](https://img.shields.io/badge/platform-macOS%2010.11-blueviolet)](https://github.com/steve228uk/YouTube-Music/releases) [![GitHub license](https://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/steve228uk/YouTube-Music/master/LICENSE)

</span>

A simple Mac app wrapper using `WKWebView` for [YouTube Music](https://music.youtube.com) that allows YouTube Music to run as a standalone process. 

## Features

- Media Keys
- Keyboard shortcuts
- "Now Playing" Notification Center widget
- Touch Bar media controls and scrubber
- System notifications when a new track plays

## Installation

You can download the latest version at the following link:

[Download Latest Version](https://github.com/steve228uk/YouTube-Music/releases/latest)

The latest binary can also be found on the releases page or clone and compile in Xcode.

### Homebrew

If [homebrew](https://brew.sh) is installed, you can also use `brew install --cask yt-music`

## Screenshots

<details><summary>View More Screenshots</summary>
<p>


![Screenshot 2](screenshots/screenshot-2.png)
![Screenshot 3](screenshots/screenshot-3.png)
![Screenshot 4](screenshots/screenshot-4.png)

</p>
</details>

## Contributing

More help in improving YouTube Music is always welcome! Please take a look at the [Issues](https://github.com/steve228uk/YouTube-Music/issues) and if there's something you think you can do, please consider filing a [pull request](https://github.com/steve228uk/YouTube-Music/pulls).

### Requirements for Building

* [Xcode](http://developer.apple.com/download/more/) - To build the project
* [CocoaPods](https://cocoapods.org) - To download all of the dependencies
* An [Apple Developer Account](https://developer.apple.com) - To sign debug builds of the app.

### Building YouTube Music

1. Clone a copy of this repo.
2. In Terminal, `cd` to the repo directory and run `pod install`.
3. Open `YT Music.xcworkspace`.
4. Under the 'Code Signing' settings in Xcode, make sure your developer account is set for Debug builds.

## Legal

Disclaimer: YT Music is an unofficial application and not affiliated with YouTube or Google Inc. in any way. "YouTube", "YouTube Music" and the "YouTube Logo" are registered trademarks of Google Inc.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

* Added preferences pane to optionally disable the global media hotkeys. [#170](https://github.com/steve228uk/YouTube-Music/pull/170)

## [1.3.3] - 2022-05-05

### Fixed

* Added an explicit check to clear out registered service workers that cause blank screens on launch. [#167](https://github.com/steve228uk/YouTube-Music/pull/167)

## [1.3.2] - 2022-05-03

### Fixed

* A periodic inconsistency where the app launches to a blank screen. [#164](https://github.com/steve228uk/YouTube-Music/pull/164)

## [1.3.1] - 2022-05-02

### Fixed

* A regression where increasing the height of the draggable toolbar was interfering with HTML elements in the page. [#162](https://github.com/steve228uk/YouTube-Music/pull/162)

## [1.3.0] - 2022-04-12

### Added

* A new application icon matching the new design language of macOS Big Sur. [#160](https://github.com/steve228uk/YouTube-Music/pull/160)
* A new keyboard shortcut for search (Command + F) to allow faster access to the search bar. [#151](https://github.com/steve228uk/YouTube-Music/pull/151)

### Fixes

* Removed the `keyDown` keyboard commaned to allow proper traversal of search results. [#151](https://github.com/steve228uk/YouTube-Music/pull/151)
* Increased navigation bar height to 64 to allow proper mouse-click passthrough. [#151](https://github.com/steve228uk/YouTube-Music/pull/151)

## [1.2.0] - 2021-10-27

### Added

* Apple Silicon support. [#149](https://github.com/steve228uk/YouTube-Music/pull/149)
* Deeply integrated support for Touch Bar on MacBook Pro. [#146](https://github.com/steve228uk/YouTube-Music/pull/146)
* Alternative global shortcut keys for controlling music playback. [#140](https://github.com/steve228uk/YouTube-Music/pull/140)

## [1.1.1] - 2021-04-27

### Fixes

* Media control keys not responding on newer versions of macOS. [#57](https://github.com/steve228uk/YouTube-Music/pull/57)
* Content not appearing in macOS Big Sur. [#130](https://github.com/steve228uk/YouTube-Music/pull/130)

## [1.1.0] - 2021-04-24

G'day everyone! Tim here! It took a fair few weeks of R&D (including [a PR to fastlane](https://github.com/fastlane/fastlane/pull/18496) itself!), but with Steve's blessing, I've finally managed to remove DevMateKit and replace it with an automated build system around GitHub Actions that will enable creating release updates of YT Music *far* more frequently than before.

This is just a basic release incorporating all of the previous improvements, but now this system in place, hopefully we can start bringing new features and OS support to the app moving forward! Enjoy!

### Added

* A Github Actions automated release pipeline. [#126](https://github.com/steve228uk/YouTube-Music/pull/126)

### Changes

* Set `NSRequiresAquaSystemAppearance` to `false` to enable Dark Mode appearance [#102](https://github.com/steve228uk/YouTube-Music/pull/102)

### Fixes

* Music auto-plays when receiving call. [#58](https://github.com/steve228uk/YouTube-Music/pull/58)
* Enabled media controller commands. [#81](https://github.com/steve228uk/YouTube-Music/pull/81)
* Latest album art being pulled correctly for every new notification. [#84](https://github.com/steve228uk/YouTube-Music/pull/84)
* Too many notifications appearing too often. [#85](https://github.com/steve228uk/YouTube-Music/pull/85)
* App crashes when resizing the window. [#87](https://github.com/steve228uk/YouTube-Music/pull/87)

## [1.0.6] - 2018-07-11

I've changed a few things in this release and it now targets 10.11 for those of you still rocking El Capitan (though you really should think about upgrading at some point 🤭).

### Media Center

This version also adds support for the native media center added in macOS 10.12.2. This means that the Today widget in Notification Center and the media center on Touch Bar devices now fully works!

### Bug Fixes

I've also fixed a bug with the like and dislike shortcuts not always liking the currently playing track.

### Feedback

As always, please leave any feedback or issues on [GitHub](https://github.com/steve228uk/youtube-music) and you can also [follow me on Twitter](https://twitter.com/steve228uk).

## [1.0.5] - 2018-07-06

Hello everyone! This update adds a couple of new features that hopefully make the YT Music experience feel a little more native.

### Keyboard Shortcuts

I've added a number of keyboard shortcuts which can be found in the new "Playback" menu in the Menu Bar. Things like Play/Pause, Next, Previous, Shuffle, Repeat, Like, and Dislike can be found here.

These are also available in the menu on the Dock icon. Right-click or option-click the YT Music icon to find these.

### Notifications

Okay, not everyone will want these and they are on by default. Like any Mac app they can be disabled in System Preferences > Notifications.

For those of you that do want these, you'll get a little banner slide in whenever the track changes. Click the banner to open YT Music or swipe it away with your mouse.

### Feedback

I hope the changes are well received and any feedback or bugs can be reported on Github

## [1.0.4] - 2018-07-02

* Only allows the window to be moved in the toolbar area at the top of the screen.

## [1.0.3] - 2018-06-19

* Added navigation arrows

## [1.0.2] - 2018-06-19

* Fixed issues with windows not displaying correctly

## [1.0.1] - 2018-06-19

* Fixes issue where window wouldn't show on first run
* Added media key support

## [1.0.0] - 2018-06-19

* Initial version of YT Music wrapper

[Unreleased]: https://github.com/steve228uk/YouTube-Music/compare/1.3.3...HEAD
[1.3.3]: https://github.com/steve228uk/YouTube-Music/compare/1.3.2...1.3.3
[1.3.2]: https://github.com/steve228uk/YouTube-Music/compare/1.3.1...1.3.2

YT Music/AppDelegate.swift
//
//  AppDelegate.swift
//  YT Music
//
//  Created by Stephen Radford on 19/06/2018.
//  Copyright © 2018 Cocoon Development Ltd. All rights reserved.
//

import Cocoa

@NSApplicationMain
class AppDelegate: NSObject, NSApplicationDelegate {
    
    lazy var mainWindowController: NSWindowController? = {
        let storyboard = NSStoryboard(name: "Main", bundle: nil)
        return storyboard.instantiateController(withIdentifier: "main") as? NSWindowController
    }()
    
    lazy var dockMenu: NSMenu = {
        let menu = NSMenu()
        menu.addItem(NSMenuItem(title: "Play", action: #selector(playPause), keyEquivalent: ""))
        menu.addItem(NSMenuItem.separator())
        menu.addItem(NSMenuItem(title: "Next", action: #selector(nextTrack), keyEquivalent: ""))
        menu.addItem(NSMenuItem(title: "Previous", action: #selector(previousTrack), keyEquivalent: ""))
        menu.addItem(NSMenuItem.separator())
        menu.addItem(NSMenuItem(title: "Shuffle", action: #selector(shuffleTracks), keyEquivalent: ""))
        menu.addItem(NSMenuItem(title: "Repeat", action: #selector(repeatTracks), keyEquivalent: ""))
        menu.addItem(NSMenuItem.separator())
        menu.addItem(NSMenuItem(title: "Like", action: #selector(likeTrack), keyEquivalent: ""))
        menu.addItem(NSMenuItem(title: "Dislike", action: #selector(dislikeTrack), keyEquivalent: ""))
        return menu
    }()
    
    func applicationDidFinishLaunching(_ notification: Notification) {
        mainWindowController?.window?.isExcludedFromWindowsMenu = true
        mainWindowController?.showWindow(self)
        mainWindowController?.window?.makeKeyAndOrderFront(self)
        NSApplication.shared.isAutomaticCustomizeTouchBarMenuItemEnabled = true
        TouchBarController.shared.setupControlStripPresence()
    }

    func applicationShouldHandleReopen(_ sender: NSApplication, hasVisibleWindows flag: Bool) -> Bool {
        if !flag {
            mainWindowController?.showWindow(self)
            mainWindowController?.window?.makeKeyAndOrderFront(self)
        }
        
        return true
    }
    
    func applicationDockMenu(_ sender: NSApplication) -> NSMenu? {
        return dockMenu
    }
    
    // MARK: - Actions
    // These are here because if the window is closed `ViewController` can't be first responder :(
    
    @objc func playPause(_ sender: Any?) {
        mainWindowController?.window?.contentViewController?
            .performSelector(onMainThread: #selector(ViewController.playPause), with: nil, waitUntilDone: true)
    }
    
    @objc func nextTrack(_ sender: Any?) {
        mainWindowController?.window?.contentViewController?
            .performSelector(onMainThread: #selector(ViewController.nextTrack), with: nil, waitUntilDone: true)
    }
    
    @objc func previousTrack(_ sender: Any?) {
        mainWindowController?.window?.contentViewController?
            .performSelector(onMainThread: #selector(ViewController.previousTrack), with: nil, waitUntilDone: true)
    }
    
    @objc func likeTrack(_ sender: Any?) {
        mainWindowController?.window?.contentViewController?
            .performSelector(onMainThread: #selector(ViewController.likeTrack), with: nil, waitUntilDone: true)
    }
    
    @objc func dislikeTrack(_ sender: Any?) {
        mainWindowController?.window?.contentViewController?
            .performSelector(onMainThread: #selector(ViewController.dislikeTrack), with: nil, waitUntilDone: true)
    }
    
    @objc func repeatTracks(_ sender: Any?) {
        mainWindowController?.window?.contentViewController?
            .performSelector(onMainThread: #selector(ViewController.repeatTracks), with: nil, waitUntilDone: true)
    }
    
    @objc func shuffleTracks(_ sender: Any?) {
        mainWindowController?.window?.contentViewController?
            .performSelector(onMainThread: #selector(ViewController.shuffleTracks), with: nil, waitUntilDone: true)
    }
    
    @IBAction func startSearch(_ sender: Any) {
        mainWindowController?.window?.contentViewController?
            .performSelector(onMainThread: #selector(ViewController.startSearch), with: nil, waitUntilDone: true)
    }
}

